pep-0544.txt

```diff
@@ -0,0 +1,968 @@
+
+* Zope interfaces [zope-interfaces]_ was one of the first widely used
+  approaches to structural subtyping in Python. It is implemented by providing
+  special classes to distinguish interface classes from normal classes,
+  to mark interface attributes, and to explicitly declare implementation.
+  For example::
+
+    from zope.interface import Interface, Attribute, implements
+
+    class IEmployee(Interface):
+
+        name = Attribute("Name of employee")
+
+        def do(work):
+            """Do some work"""
+
+    class Employee(object):
+        implements(IEmployee)
+
+        name = 'Anonymous'
+
+        def do(self, work):
+            return work.start()
+
+  Zope interfaces support various contracts and constraints for interface
+  classes. For example::
+
+    from zope.interface import invariant
+
+    def required_contact(obj):
+        if not (obj.email or obj.phone):
+            raise Exception("At least one contact info is required")
+
+    class IPerson(Interface):
+
+        name = Attribute("Name")
+        email = Attribute("Email Address")
+        phone = Attribute("Phone Number")
+
+        invariant(required_contact)
+
+  Even more detailed invariants are supported. However, Zope interfaces rely
+  entirely on runtime validation. Such focus on runtime properties goes
+  beyond the scope of the current proposal, and static support for invariants
+  might be difficult to implement. However, the idea of marking an interface
+  class with a special base class is reasonable and easy to implement both
+  statically and at runtime.
+
+* Python abstract base classes [abstract-classes]_ are the standard
+  library tool to provide some functionality similar to structural subtyping.
+  The drawback of this approach is the necessity to either subclass
+  the abstract class or register an implementation explicitly::
+
+    from abc import ABC
+
+    class MyTuple(ABC):
+        pass
+
+    MyTuple.register(tuple)
+
+    assert issubclass(tuple, MyTuple)
+    assert isinstance((), MyTuple)
+
+  As mentioned in the `rationale`_, we want to avoid such necessity, especially
+  in static context. However, in a runtime context, ABCs are good candidates for
+  protocol classes and they are already used extensively in
+  the ``typing`` module.
+
+* Abstract classes defined in ``collections.abc`` module [collections-abc]_
+  are slightly more advanced since they implement a custom
+  ``__subclasshook__()`` method that allows runtime structural checks without
+  explicit registration::
+
+    from collections.abc import Iterable
+
+    class MyIterable:
+        def __iter__(self):
+            return []
+
+    assert isinstance(MyIterable(), Iterable)
+
+  Such behavior seems to be a perfect fit for both runtime and static behavior
+  of protocols. The main goal of this proposal is to support such behavior
+  statically. In addition, to allow users achieving such runtime behavior
```

Comment:
http://ell.stackexchange.com/questions/11193/allow-to-infinitive-substantive-verb-ing says it probably should be "to achieve" here.